text/0000-iterator-try_all-try_any.md

```diff
@@ -0,0 +1,144 @@
+- Feature Name: iter_try_all_any
+- Start Date: 2022-02-16
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+
+[summary]: #summary
+
+Add the associated methods `try_any` and `try_all` to the `Iterator` trait, which work similar to `any` and `all`, but accept `Result` and `Option` as closure return type.
+
+# Motivation
+
+[motivation]: #motivation
+
+In some cases where one would like to use `any`/`all`, the actual condition to check is fallible. For example, let's say we have a simple key-value-store like a `HashMap`. The keys are IDs to the objects. We further have a list of IDs we want to check some condition on. We would like to use `all` on an iterator over the list to ensure all objects match our condition. However, access to elements is fallable and returns `Option`s. In other cases, the checks might involve more complex validation on the objects. The validation method could involve access to external parts and fail. Therefore, it returns a `Result<bool, Error>`.
+
+The implementor needs to return a `bool` for every element in `all`. They could use `.try_fold(true, |state, item| Ok(state && check(item)?))`, but this is less readable and does not short-circuit as `all` does. If they would do `.all(|item| check(item).unwrap_or(false))`, the error would be discarded and lost/ignored. To fully comply with expectations, there is only two possibilites (I see):
```

Comment:
Short-circuiting on both is probably easiest done with `try_for_each`, something like
```rust
    let r = it.try_for_each(|x| match check(x) {
        Ok(b) => if b { ControlFlow::Continue(()) } else { ControlFlow::Break(Ok(false)) },
        e => ControlFlow::Break(e),
    });
    let r = match r {
        ControlFlow::Continue(()) => Ok(true),
        ControlFlow::Break(r) => r,
    };
```
<https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=746bd31e2418ffacdff08855c91e6205>

But yes, that's of course much less readable than having the method provided by `Iterator`.

Reply:
Oh interesting, thanks, that is another possibility indeed. Not a good one of course, I would expect anyone to go for another route.